ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -2226,25 +2229,286 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
+
 	switch (get_class(*head)) {
 	case PKCS11_CKO_SECRET_KEY:
 		return set_secret_key_data(head, data, key_size);
+	case PKCS11_CKO_PRIVATE_KEY:
+		type = get_key_type(*head);
+		if (type == PKCS11_CKK_RSA)
+			return set_private_key_data_rsa(head, data, key_size);
+		break;
 	default:
 		return PKCS11_CKR_GENERAL_ERROR;
 	}
+
+	return PKCS11_CKR_GENERAL_ERROR;
+}
+
+static enum pkcs11_rc
+encode_rsa_private_key_der(struct obj_attrs *head, void **data, uint32_t *sz)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	int i = 0;
+	int mbedtls_rc = 0;
+	int start = 0;
+	int der_size = 0;
+	void *N = NULL;
+	void *P = NULL;
+	void *Q = NULL;
+	void *D = NULL;
+	void *E = NULL;
+	uint32_t N_len = 0;
+	uint32_t P_len = 0;
+	uint32_t Q_len = 0;
+	uint32_t D_len = 0;
+	uint32_t E_len = 0;
+	uint8_t *buffer = NULL;
+	mbedtls_pk_context pk;
+	mbedtls_rsa_context *rsa;
+	const mbedtls_pk_info_t *pk_info;
+
+	mbedtls_pk_init(&pk);
+	pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_RSA);
+	if (mbedtls_pk_setup(&pk, pk_info)) {
+		rc = PKCS11_CKR_GENERAL_ERROR;
+		goto out;
+	}
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_MODULUS, &N, &N_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PRIME_1, &P, &P_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PRIME_2, &Q, &Q_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PRIVATE_EXPONENT, &D, &D_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PUBLIC_EXPONENT, &E, &E_len);
+	if (rc)
+		goto out;
+
+	rsa = mbedtls_pk_rsa(pk);
+	mbedtls_rc = mbedtls_rsa_import_raw(rsa, N, N_len, P, P_len,
+					    Q, Q_len, D, D_len, E, E_len);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (mbedtls_rsa_complete(rsa)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (mbedtls_rsa_check_privkey(rsa)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	der_size = N_len * 8;
+	buffer = TEE_Malloc(der_size, TEE_MALLOC_FILL_ZERO);
+	if (!buffer) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_pk_write_key_der(&pk, buffer, der_size);
+	if (mbedtls_rc < 0) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	start = der_size - mbedtls_rc;
+	for (i = 0; i < mbedtls_rc; i++) {
+		buffer[i] = buffer[i + start];
+		buffer[i + start] = 0;
+	}
+
+	*data = buffer;
+	*sz = mbedtls_rc;
+out:
+	mbedtls_pk_free(&pk);
+	if (rc) {
+		TEE_Free(buffer);
+		*data = NULL;
+		*sz = 0;
+	}
+
+	return rc;
 }
 
 enum pkcs11_rc get_key_data_to_wrap(struct obj_attrs *head, void **data,
 				    uint32_t *sz)
 {
+	uint32_t type = 0;
 	switch (get_class(head)) {
 	case PKCS11_CKO_SECRET_KEY:
 		if (get_attribute_ptr(head, PKCS11_CKA_VALUE, data, sz))
 			return PKCS11_CKR_ARGUMENTS_BAD;
 		break;
+	case PKCS11_CKO_PRIVATE_KEY:
+		type = get_key_type(head);
+		if (type == PKCS11_CKK_RSA)
+			return encode_rsa_private_key_der(head, data, sz);
+		return PKCS11_CKR_GENERAL_ERROR;
```

Comment:
```suggestion
		fallthrough;
```
to get rid of repetitive returns

Reply:
That will result in compilation warning

ta/pkcs11/src/pkcs11_attributes.c: In function ‘get_key_data_to_wrap’:
ta/pkcs11/src/pkcs11_attributes.c:2513:6: warning: this statement may fall through [-Wimplicit-fallthrough=]

to get rid of repetitive returns
set rc value instead of return
https://github.com/OP-TEE/optee_os/pull/5647/commits/fcdfab34fd6a9f66fe3809e644fd15da69a8dfcf